src/device/ping-device.cpp

```diff
@@ -0,0 +1,100 @@
+#include "ping-device.h"
+
+#include <time/ping-time.h>
+#include <link/ping-port.h>
+
+#include <cstdio>
+
+bool PingDevice::initialize()
+{
+    return request(CommonId::PROTOCOL_VERSION) && request(CommonId::DEVICE_INFORMATION);
+}
+
+ping_message* PingDevice::read()
+{
+    uint8_t data;
+    int result = _port.read(&data, 1);
+    if (result != 0 && _parser.parseByte(data) == PingParser::NEW_MESSAGE) {
+        return &_parser.rxMessage;
```

Comment:
can't we return a copy of the data? I usually get scared with returning pointers to temporaries. There's no hard checking that the poitner to rxMessage would not be sued after _parser dies.

Reply:
Check my comments in the PR body, there is the explanation for that and the alternative for that where you can get a copy.